site/blog/2022-11-04.mdx

```diff
@@ -0,0 +1,123 @@
+
+## ** High-level vs Low-level IaSQL Modules **
+In IaSQL, you are able to do cloud actions both on a high-level and a low-level abstraction layer. It means that you can go with both of these two approaches:
+1. Deploy the folder `examples/ecs-fargate/prisma/app` of the repo `https://github.com/iasql/iasql-engine/` to ECS on my AWS account
+2. Create a security group named `allow-all` attached to the VPC with ID `vpc-f56cff91`. Then add a rule to it allowing all incoming traffic on port `80` from `0.0.0.0/0`.
+
+The first statement, is described in a high-level abstraction layer. The higher the level, the less detail. So it doesn't say anything about the underlying steps that should be taken in order for the statement to be executed. But the second statement is giving all the details necessary to create a new security group.
+
+The flexibility of IaSQL– which makes it really awesome– is that you can create both the high-level and the low-level modules for it. Both of the above statements are already possible using IaSQL modules. The first one using a module named `aws_ecs_simplified` and the latter using the `aws_security_group` module. Let's take a peek on the actual implementations really quick before we continue our main topic.
+
+```sql title="Deploy the folder 'examples/ecs-fargate/prisma/app' of the 'iasql-engine' repo to ECS on my AWS account"
+-- Install the "high-level" aws_ecs_simplified IaSQL module
+SELECT iasql_install('aws_ecs_simplified');
+-- Create the load balancer, ECS service, task definition, etc. All ready afterwards for us to build and push to deploy the app.
+INSERT INTO ecs_simplified (app_name, app_port, image_tag, public_ip) VALUES ('simple-express', 8088, 'latest', true);
+SELECT iasql_apply();
+-- Use the "high-level" ecr_build function to build the folder "examples/ecs-fargate/prisma/app" from "iasql-engine" Github repo and push it to ECR
+SELECT ecr_build(
+  'https://github.com/iasql/iasql-engine/',
+  (SELECT id FROM repository WHERE repository_name = 'simple-express-repository')::varchar(255),
+  './examples/ecs-fargate/prisma/app',
+  'ghp_XXX', -- replace your github personal access token here
+  'main'
+);
+```
+```sql title="Create a security group named 'allow-all' attached to the VPC with ID 'vpc-f56cff91'. Then add a rule to it allowing all incoming traffic on port '80' from '0.0.0.0/0'"
+-- Install the aws_security_group IaSQL module
+SELECT iasql_install('aws_security_group');
+-- "Create a security group named `allow-all` attached to the VPC with ID `vpc-f56cff91`"
+INSERT INTO security_group (description, group_name, vpc_id)
+VALUES ('allows all incoming traffic on port 80', 'allow-all', (SELECT id FROM vpc WHERE vpc_id = 'vpc-08c0a68025c2983a2'));
```

Comment:
You have two different `vpc_id`s referenced in this text.

I would also suggest restructuring this to select the `vpc_id` through the [`subnet`s on the `service` record](https://github.com/iasql/iasql-engine/blob/main/src/modules/0.0.22/aws_ecs_fargate/entity/service.ts#L67).

Something like:

```suggestion
-- "Create a security group named `allow-all` attached to the VPC the ECS service is on"
INSERT INTO security_group (description, group_name, vpc_id)
VALUES ('allows all incoming traffic on port 80', 'allow-all', (
  SELECT id
  FROM vpc
  INNER JOIN subnet ON subnet.vpc_id = vpc.id
  INNER JOIN service ON service.subnets[1] = subnet.subnet_id
  WHERE service.name = 'simple-express-service'
 ));
```

It's a bit longer, but much less error prone, and truly runnable with the Button links without requiring any edits on the user's side, if they want.

Reply:
This example didn't meant to be related to the `ecs_simplified` context. I was trying to tell the users that they can do both the high-level and low-level stuff using IaSQL. But if it's not self-explanatory enough, I should add some more info showing that intention?

Reply:
Then I am completely confused by the example being in there, to be honest? It feels very off-topic from the "deploy my `Dockerfile`-managed git repo to AWS easily" subject if it is not used in the actual flow in any way, and could similarly confuse people into believing it is necessary.

If you want to go into the high-level/low-level distinctions in this blog post, perhaps it would make more sense explaining how `aws_ecs_simplified` depends on `aws_ecs_fargate`, `aws_ecr`, etc, and listing out all of the SQL statements it's `INSERT` trigger executes for you? That would showcase how *much* the simplified module does for you.

Then you could take it further, if you wanted to, showing how after you created the deployment via the simplified module you can still `UPDATE` some default that you have a special need to be different, getting the power to adjust things to your precise needs that a PaaS would instead force you to work around because it cannot be changed.